### PR TITLE
Consolidate the two auxiliary tries to the main NFD trie

### LINKED ISSUE #2362: Consolidate two auxiliary tries to the main NFD trie
The trie for the canonical combining class lives in uprops due to thinking that it's a property so it belongs there. However, this results in lookups from two tries for the same character: both from the recursive decomposition trie and the canonical combining class trie. As it turns out, it's easy to bake the values of the latter into the former. This applies to all normalization forms and collation.

The trie for composing normalization passthrough optimization holds data that could be deduced from the decomposition trie and (when applicable) its supplement if simple marker values were added to the decomposition trie for a couple of cases.

With these changes, three trie lookups could be collapsed into one, which seems good for performance. It would be sad not to do this. However, to do this for 1.0, the normalization data format needs to change between 1.0-beta and 1.0.

Additionally, with this change, for any app that uses any of `Collator`, `CanonicalDecomposition`, `DecomposingNormalizer`, or `ComposingNormalizer`, the canonical combining class data in uprops is duplicate data. OTOH, if an app somehow managed to be interested in the canonical combining class without being interested in the APIs from the previous sentence, the trie in uprops would be smaller than bringing in the whole decomposition trie. But that's not particularly realistic. E.g. HarfBuzz wants also `CanonicalDecomposition`.

Note that this not only consolidates data but also the number of lookup actions at run time.

### PR DESCRIPTION
**Pull request by hsivonen**
Closes #2362.

Improves throughput from +15% (Greek) to +119% (English) even with the iterator formulation.

GitHub says -6.42 KB for testdata.postcard.

### CODE REVIEW

**Comment by hsivonen** [comment]
* [ICU4C PR to main](https://github.com/unicode-org/icu/pull/2152)
* [ICU4C PR to maint-71](https://github.com/unicode-org/icu/pull/2153)
* [ICU4C issue](https://unicode-org.atlassian.net/browse/ICU-22115)

File: components/collator/src/elements.rs [hunk 1]
```diff
@@ -623,8 +670,13 @@ impl Default for NonPrimary {
 /// when `upcoming[0]` is an undecomposed starter, we don't
 /// need the ccc yet, and when lookahead has already done the
 /// trie lookups, we don't need `trie_value`, as it is implied
-/// by ccc. Optimizing size by this observation seems
-/// unnecessary, since `upcoming` is typically very short.
+/// by ccc.
+///
+/// TODO: This struct carries redundant information, and
```
Comment (echeran) [inline components/collator/src/elements.rs:675]: nit (optional): create issue and update as `TODO(#issue)`

File: components/normalizer/src/lib.rs [hunk 2]
```diff
@@ -629,32 +676,37 @@ where
         debug_assert!(self.pending.is_none());
         let c = self.delegate.next()?;
 
-        if self.has_starter_exceptions {
-            if u32::from(c) & !1 == 0xFF9E && self.half_width_voicing_marks_become_non_starters {
-                return Some(CharacterAndTrieValue {
-                    character: c,
-                    trie_val: DECOMPOSITION_STARTS_WITH_NON_STARTER,
-                });
+        // TODO: Measure if this check is actually an optimization even in the
```
Comment (echeran) [inline components/normalizer/src/lib.rs:679]: ditto

File: components/normalizer/src/properties.rs [hunk 3]
```diff
@@ -0,0 +1,402 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! Access to the Unicode properties or property-based operations that
+//! are required for NFC and NFD.
+//!
+//! Applications should generally use the full normalizers that are
+//! provided at the top level of this crate. However, the APIs in this
+//! module are provided for callers such as HarfBuzz that specifically
```
Comment (echeran) [inline components/normalizer/src/properties.rs:10]:
```suggestion
//! module are provided for callers such as HarfBuzz that specifically
//! want access to the raw canonical composition operation e.g. for use in a
//! glyph-availability-guided custom normalizer.
```

File: components/normalizer/src/properties.rs [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+
+impl CanonicalCombiningClassMap {
+    /// Look up the canonical combining class for a scalar value
+    #[inline(always)]
+    pub fn get(&self, c: char) -> CanonicalCombiningClass {
+        self.get_u32(u32::from(c))
+    }
+
+    /// Look up the canonical combining class for a scalar value
+    /// represented as `u32`. If the argument is outside the scaler
```
Comment (echeran) [inline components/normalizer/src/properties.rs:375]:
```suggestion
    /// represented as `u32`. If the argument is outside the scalar
```

**Review by echeran — APPROVED** [review]
Looks good.  I added very minor suggestions. Feel free to merge and then follow up later on filing issues to match & link to the unadorned TODOs.

**Comment by echeran** [comment]
> * [ICU4C PR to main](https://github.com/unicode-org/icu/pull/2152)
> * [ICU4C PR to maint-71](https://github.com/unicode-org/icu/pull/2153)
> * [ICU4C issue](https://unicode-org.atlassian.net/browse/ICU-22115)

Thanks for doing this. FYI, it sounds like in the future, we can just open one ticket per ICU release for ICU4X export data tool updates.  That would allow us to reuse a ticket for multiple PRs within an ICU release without having to explicitly get TC approval for each subsequent PR.

File: components/collator/src/elements.rs (hunk 1 above)
Comment (hsivonen) [reply components/collator/src/elements.rs]: Done

File: components/normalizer/src/lib.rs (hunk 2 above)
Comment (hsivonen) [reply components/normalizer/src/lib.rs]: Done

**Comment by hsivonen** [comment]
Thanks for the review. I've addressed the review comments here. I still intend to push one more changeset with a CI job update to full datagen zip once the ICU4C change is tagged for export from the maint-71 branch (by @sffc like last time?).

**Comment by hsivonen** [comment]
Thank you for the icuexportdata update. This still needs re-approval for GitHub to permit landing.

**Review by echeran — APPROVED** [review]

**Comment by hsivonen** [comment]
@echeran , sorry about requesting review yet again. A conflicting change landed on `main`. Could you, please, press the landing button upon approval so that there isn't a time window for another merge conflict?

**Review by echeran — APPROVED** [review]
Sorry about the conflicts.  Yeah, I should have just merged given the volume of things right before 1.0.  I'll merge now.

**Pull request merged**